### PR TITLE
UI: improve Dockerfile for faster build

### LINKED ISSUE #1139: Optimize UI container build
**Component**: ui

**Why this is needed**: The UI container is extremely slow to build (~15 minutes). This should be remedied now rather than later.

**What should be done**: Make the UI container build fast again (MTUICBFA?).

**Implementation proposal** (strongly recommended): Most likely some optimization to be done around docker file.

**Test plan**:

### PR DESCRIPTION
**Pull request by carlito-scality**
**Component**: UI

**Context**: 
The UI is extremely slow to build (~15 minutes). This should be remedied now rather than later.

**Summary**:
Some docker layers could be cached.

**Acceptance criteria**: 
The UI build is faster after the first build since some docker layers are cached.
Limitations: when rebuilding UI, we have to restart the pods manually to pull the new UI image (correctly generated) since the UI version does not change every times. Otherwise, the new image will be not correctly deployed.
``` kubectl delete pods metalk8s-ui-xxx-xxx -n kube-system ```

Closes: https://github.com/scality/metalk8s/issues/1139

### CODE REVIEW

**Review by gdemonet — DISMISSED** [review]
LGTM, great speed improvements!

File: ui/Dockerfile [hunk 1]
```diff
@@ -3,8 +3,11 @@ ARG NGINX_IMAGE_VERSION=1.15.8
 FROM node AS build-step
 
 WORKDIR /home/node
-COPY . /home/node/
+COPY package*.json /home/node/
```
Comment (slaperche-scality) [inline ui/Dockerfile:7]: You should also update the `file_dep` for the `targets.LocalImage` for the UI image in the buildchain, no?

File: ui/Dockerfile [hunk 2]
```diff
@@ -1,6 +1,7 @@
 ARG NGINX_IMAGE_VERSION=1.15.8
+ARG NODE_IMAGE_VERSION=10.16.0
 
-FROM node AS build-step
+FROM node:${NODE_IMAGE_VERSION} AS build-step
```
Comment (slaperche-scality) [inline ui/Dockerfile:4]: The value for this variable comes from where?

Comment (carlito-scality) [reply ui/Dockerfile:4]: ARG NODE_IMAGE_VERSION=10.16.0

Comment (carlito-scality) [reply ui/Dockerfile:4]: I will put it in targets.LocalImage of image.py as well

File: ui/Dockerfile (hunk 1 above)
Comment (carlito-scality) [reply ui/Dockerfile:7]: I think we do not have to update since package*.json is part of '/ui'.

Comment (slaperche-scality) [reply ui/Dockerfile:7]: Yeah, but currently the buildchain will re-trigger the build if any file under `/ui` is modified, whereas with you change we only care about the `package*.json`.

If you don't update then you'll trigger useless builds (that will be fast thanks to better caching with this change), so it's better to update to avoid it :)

BTW, shouldn't your check complains about this mismatch @jbertran ?

File: buildchain/buildchain/image.py [hunk 3]
```diff
@@ -304,11 +304,11 @@ def show() -> str:
         build_args={'NGINX_IMAGE_VERSION': NGINX_IMAGE_VERSION, 'NODE_IMAGE_VERSION': NODE_IMAGE_VERSION},
         task_dep=['_image_mkdir_root'],
         file_dep=(
```
Comment (slaperche-scality) [inline buildchain/buildchain/image.py:309]: This commit should be squashed with the previous one.

File: buildchain/buildchain/image.py [hunk 4]
```diff
@@ -301,7 +301,7 @@ def show() -> str:
         dockerfile=constants.ROOT/'ui'/'Dockerfile',
         destination=constants.ISO_IMAGE_ROOT,
         save_on_disk=True,
-        build_args={'NGINX_IMAGE_VERSION': NGINX_IMAGE_VERSION},
+        build_args={'NGINX_IMAGE_VERSION': NGINX_IMAGE_VERSION, 'NODE_IMAGE_VERSION': NODE_IMAGE_VERSION},
```
Comment (slaperche-scality) [inline buildchain/buildchain/image.py]: line too long, please wrap it as:

```
        build_args={
            'NGINX_IMAGE_VERSION': NGINX_IMAGE_VERSION,
            'NODE_IMAGE_VERSION': NODE_IMAGE_VERSION
        },
```

File: salt/metalk8s/addons/ui/files/metalk8s-ui-deployment.yaml [hunk 5]
```diff
@@ -27,6 +27,7 @@ spec:
       containers:
         - name: metalk8s-ui
           image: {{ build_image_name('metalk8s-ui', '0.2') }}
+          imagePullPolicy: Always
```
Comment (slaperche-scality) [inline salt/metalk8s/addons/ui/files/metalk8s-ui-deployment.yaml:30]: unrelated change, shouldn't be in this commit

File: ui/src/containers/Layout.js [hunk 6]
```diff
@@ -99,7 +99,6 @@ class Layout extends Component {
             <PrivateRoute exact path="/about" component={Welcome} />
             <PrivateRoute exact path="/" component={NodeList} />
           </Switch>
-          )}
```
Comment (slaperche-scality) [inline ui/src/containers/Layout.js:102]: unrelated change, shouldn't be in this commit

**Review by slaperche-scality — CHANGES_REQUESTED** [review]

File: salt/metalk8s/addons/ui/files/metalk8s-ui-deployment.yaml (hunk 5 above)
Comment (carlito-scality) [reply salt/metalk8s/addons/ui/files/metalk8s-ui-deployment.yaml:30]: I rename the commit in the squash

File: ui/src/containers/Layout.js (hunk 6 above)
Comment (carlito-scality) [reply ui/src/containers/Layout.js:102]: I rename the commit in the squash

File: buildchain/buildchain/image.py (hunk 3 above)
Comment (carlito-scality) [reply buildchain/buildchain/image.py:309]: done

**Comment by carlito-scality** [comment]
@bert-e reset

**Comment by carlito-scality** [comment]
@bert-e reset

File: buildchain/buildchain/image.py [hunk 7]
```diff
@@ -87,7 +87,7 @@ def show() -> str:
 
 
 NGINX_IMAGE_VERSION : str = '1.15.8'
-
+NODE_IMAGE_VERSION : str = '10.16.0'
```
Comment (slaperche-scality) [inline buildchain/buildchain/image.py:90]:
```
-update file_dep
-change the pull policy deployment.yaml
-remove useless html element
```

Those 3 changes are totally unrelated (unless I miss something) and should have dedicated commits.

A commit is "atomic" and has one goal: what if I want to revert/cherry-pick the removal of the useless HTML element without losing/picking the `file_dep` update?

**Review by slaperche-scality — CHANGES_REQUESTED** [review]

File: buildchain/buildchain/image.py (hunk 7 above)
Comment (carlito-scality) [reply buildchain/buildchain/image.py:90]: Two first changes are part of the build configuration. Ok the last one is a missing error but I will not create another for 2 characters removal.

Comment (slaperche-scality) [reply buildchain/buildchain/image.py:90]: the deployment.yaml is not related to the buildchain, but to the deployment so they don't belong to the same commit.

Having a dedicated commit for this one would allow you to explains why you added ` imagePullPolicy: Always` in the commit message (useful when someone will use `git log` or `git blame`).

And you should create a dedicated commit, even for your two characters removal since it has absolutely no relation with the rest (a little commit "fix typo" cost nothing and keep a usable and useful git history, which save a lot more time in the future than splitting the commit).

**Review by slaperche-scality — CHANGES_REQUESTED** [review]

File: buildchain/buildchain/image.py (hunk 7 above)
Comment (carlito-scality) [reply buildchain/buildchain/image.py:90]: Ok for the last one. but for the two first, I do not agree. It depends on how we define the "atomicity" of a commit.

**Comment by carlito-scality** [comment]
@bert-e reset

**Review by slaperche-scality — APPROVED** [review]

**Comment by carlito-scality** [comment]
@bert-e approve
